### DISCUSSION TITLE
NaN in Metropolis acceptance decision

### DISCUSSION THREAD

**Discussion by skkestrel** [netket/netket#663] [General]
Hello, 

I am running the following code (slightly modified from the examples)

```python
from netket import legacy as nk

# 1D Periodic Lattice
g = nk.graph.Hypercube(length=8, n_dim=1, pbc=True)

# Boson Hilbert Space
hi = nk.hilbert.Boson(N=g.n_nodes, n_max=3, n_bosons=None)

# Bose Hubbard Hamiltonian
ha = nk.operator.BoseHubbard(hilbert=hi, graph=g, U=8.0, mu=1.0)

# RBM Machine with one-hot encoding, real parameters, and symmetries
ma = nk.machine.RbmMultiVal(hilbert=hi, alpha=1, dtype=float, automorphisms=g)
ma.init_random_parameters(seed=1234, sigma=0.01)

# Sampler using Hamiltonian moves, thus preserving the total number of particles
sa = nk.sampler.MetropolisHamiltonian(machine=ma, hamiltonian=ha)

# Stochastic gradient descent optimization
op = nk.optimizer.Sgd(ma, 0.05)

# Variational Monte Carlo
sr = nk.optimizer.SR(ma, diag_shift=0.1)
vmc = nk.Vmc(
    hamiltonian=ha, sampler=sa, optimizer=op, n_samples=4000, n_discard=0, sr=sr
)

vmc.run(n_iter=300, out="test")
```

and I get the following error during the Metropolis update step.

 ```python
File "bosehubbard1d.py", line 31, in <module>
    vmc.run(n_iter=300, out=f"data/{8}_{4.0 * i / 8}", obs=obs)
  File "/home/kz/.local/lib/python3.8/site-packages/netket/legacy/abstract_variational_driver.py", line 231, in run
    for step in itr:
  File "/home/kz/.local/lib/python3.8/site-packages/tqdm/std.py", line 1178, in __iter__
    for obj in iterable:
  File "/home/kz/.local/lib/python3.8/site-packages/netket/legacy/abstract_variational_driver.py", line 153, in iter
    dp = self._forward_and_backward()
  File "/home/kz/.local/lib/python3.8/site-packages/netket/legacy/_vmc.py", line 144, in _forward_and_backward
    self._samples = self._sampler.generate_samples(
  File "/home/kz/.local/lib/python3.8/site-packages/netket/legacy/sampler/abstract_sampler.py", line 58, in generate_samples
    samples[i] = self.__next__()
  File "/home/kz/.local/lib/python3.8/site-packages/netket/legacy/sampler/numpy/metropolis_hastings.py", line 114, in __next__
    accepted += _acc_kernel(
  File "/home/kz/.local/lib/python3.8/site-packages/netket/legacy/sampler/numpy/metropolis_hastings.py", line 83, in acceptance_kernel
    assert not math.isnan(prob)
AssertionError
```

Can anyone advise as to how to resolve this error? It seems to be related to the parameters of the bose-hubbard Hamiltonian.

**Comment by PhilipVinc** [comment]
Hi @skkestrel 

I think the NaNs are caused by SR. You are probably taking too large steps at the begininng (maybe try lowering the learning rate?) and eventually you get nans in the parameters.

However, I'd point out that you are looking at legacy (netket v2) examples.
`netket.legacy` is there only to aid people who have old code working on previous versions of netket transit it to the new API and models of netket v3, however it will be deleted with the release of netket 3.1 in a few months, and i'd strongly discourage using it.

I'd strongly invite you to play with the new netket.
We still haven't updated all code examples (the ones in legacy, indeed) but you should be able to use the new netket to study that same model.
The only thing that isn't implemented out of the box is a RBMMultiVal with Symmetries (cc @femtobit ) but if you practice a bit with simpler models you might later try to write it yourself. It should be only 5-6 lines of code.

**Comment by VolodyaCO** [comment]
I remember having to deal with something of the sort. Indeed, what happened is that gradients began to explode and nans were introduced in the machine's predictions. I believe I dealt with that successfully by adding regularisation, to keep parameter values "sane". I think `nk.optimizer.Sgd` has a regularisation argument.

**Comment by PhilipVinc** [comment]
@skkestrel By the way, we pushed out an update fixing a few corner cases with BoseHubbard Hamiltonians. If you are going to trying netket 3 make sure to update.

@VolodyaCO No, it got lost in the `great rewrite`. It is somewhat trivial to add regularisation with [optax](https://optax.readthedocs.io/en/latest/) though. (Netket wrappers are simply optax optimisers with slightly different names to conform to our old API). It would be interesting to add back the regularisation by wrapping optax,sgd into an `optax.clip` or `optax.clip_byglobal_norm` or something

**Comment by VolodyaCO** [comment]
Yeah @PhilipVinc, I mean, if @skkestrel continues to use `netket.legacy`, the Sgd optimiser has the regularisation argument.

**Comment by PhilipVinc** [comment]
I converted the issue to a discussion as this is not about netket itself. Feel free to get back here for feedback on how to optimise the problems you are working on.

**Comment by skkestrel** [comment]
Thanks all for the help. This has been extremely informative for me. I have upgraded my code to netket 3 and I will try out just reducing the learning rate to prevent the gradients from blowing up immediately first - it seems to be working much better now.

I'd also like to ask, is there a recommended method for working inside the whole subspace (i.e. not restricted to a certain boson number)? I have firstly set n_particles=None to remove the subspace restriction. However, MetropolisHamiltonian preserves particle number. Is there another built-in sampler that is suitable for working in variable particle number subspace?

Thanks.

**Reply by PhilipVinc** [reply]
Have a look around our list of [samplers](https://www.netket.org/docs/api.html#list-of-samplers).
In general the trivial `MetropolisLocal` will flip the local state at random and won't conserve the photon number.

You can also write your own transition rule for metropolis if you want by modifying [localrule](https://www.netket.org/_modules/netket/sampler/rules/local.html#LocalRule).

**Reply by VolodyaCO** [reply]
MetropolisLocal should do the trick. You can also implement a custom sampler, because metropolislocal has reportedly shown some ergodicity-related issues. Not bad though if you want to play around with it.  Check https://www.netket.org/docs/api.html#list-of-samplers

**Reply by skkestrel** [reply]
Thanks!
